lib/portage/util/file_copy.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+import errno
+import fcntl
+import os
+import platform
+import shutil
+
+
+FICLONE = 0x40049409
+
+
+def _get_chunks(src):
+    if not hasattr(os, "SEEK_DATA"):
+        offset_end = os.lseek(src, 0, os.SEEK_END)
+        yield (0, offset_end)
+        return
+
+    offset_hole = 0
+    while True:
+        try:
+            # Find the next bit of data
+            offset_data = os.lseek(src, offset_hole, os.SEEK_DATA)
+        except OSError as e:
+            if e.errno not in (errno.EINVAL, errno.ENXIO):
+                raise
+
+            offset_end = os.lseek(src, 0, os.SEEK_END)
+
+            if e.errno == errno.EINVAL:
+                # SEEK_DATA failed, return the whole file
+                yield (0, offset_end)
+
+            elif offset_end > offset_hole:
+                # Hole at end of file
+                yield (offset_end, 0)
+
+            break
+        else:
+            offset_hole = os.lseek(src, offset_data, os.SEEK_HOLE)
+            yield (offset_data, offset_hole - offset_data)
+
+
+def _do_copy_file_range(src, dst):
+    for offset, count in _get_chunks(src):
+        if count == 0:
+            os.ftruncate(dst, offset)
+        else:
+            while count > 0:
+                written = os.copy_file_range(src, dst, count, offset, offset)
+                offset += written
+                count -= written
+
+
+def _do_sendfile(src, dst):
+    for offset, count in _get_chunks(src):
+        if count == 0:
+            os.ftruncate(dst, offset)
+        else:
+            os.lseek(dst, offset, os.SEEK_SET)
+            while count > 0:
+                written = os.sendfile(dst, src, offset, count)
+                offset += written
+                count -= written
+
+
+def _fastcopy(src, dst):
+    with (
+        open(src, "rb", buffering=0) as srcf,
+        open(dst, "wb", buffering=0) as dstf,
+    ):
+        srcfd = srcf.fileno()
+        dstfd = dstf.fileno()
+
+        if platform.system() == "Linux":
+            try:
+                fcntl.ioctl(dstfd, FICLONE, srcfd)
+                return True
+            except OSError:
+                pass
+
+        if hasattr(os, "copy_file_range"):
+            try:
+                _do_copy_file_range(srcfd, dstfd)
+                return True
+            except OSError:
+                pass
```

Comment:
With this ignorance of OSError we won't catch the zfs bugs anymore.  :smirk:

Reply:
If we do a something like this then we can still catch zfs bugs:
```python
logging.warning("_do_copy_file_range failed", exc_info=sys.exc_info())
```

Reply:
If there are specific errno values you think we should handle differently, happy to make changes here.

Reply:
If we do a something like this then we can still catch zfs bugs:
```python
logging.warning("_do_copy_file_range failed", exc_info=sys.exc_info())
```

Reply:
It will also give us some visibility to better understand _do_copy_file_range failures.

Reply:
I added warnings for copy_file_range and sendfile. Let me know what you think.

Reply:
That looks great, thanks!